approximate_string_matching/four_russians_helpers.py

```diff
@@ -0,0 +1,260 @@
+      D = max([self.delete_cost_function(letter_idx) for letter_idx in A])
+      return max(I, D)
+
+    m = get_parameter(text_1)
+    step_size_bound = get_step_size_bound()
+
+    text_1_mod = (len(text_1) - 1) % m
+    if text_1_mod != 0:
+      text_1 += '#' * (m - (text_1_mod))
+
+    text_2_mod = (len(text_2) - 1) % m
+    if text_2_mod != 0:
+      text_2 += '#' * (m - (text_2_mod))
+
+    return m, A, step_size_bound, text_1, text_2
+
+  @staticmethod
+  def get_all_strings(m, A, dumb_letter):
+    """ Returns all possible strings of a given length """
+
+    sub_length_results = [[] for _ in range(m + 1)]
+    sub_length_results[0] = []
+    sub_length_results[1] = A
+
+    for size in range(2, m + 1):
+      sub_length_results[size] = []
+      for left_substring in sub_length_results[1]:
+        for right_substring in sub_length_results[size - 1]:
+          sub_length_results[size].append(left_substring + right_substring)
+
+    result = [dumb_letter + word for word in sub_length_results[m]]
+    return result
+
+  @staticmethod
+  def store(R_new, S_new, C, D, R, S, storage):
+    R = tuple(R)
+    S = tuple(S)
+
+    if C not in storage:
+      storage[C] = {}
+    if D not in storage[C]:
+      storage[C][D] = {}
+    if R not in storage[C][D]:
+      storage[C][D][R] = {}
+
+    storage[C][D][R][S] = (R_new, S_new)
+
+  @staticmethod
+  def fetch(C, D, R, S, storage):
+    return storage[C][D][tuple(R)][tuple(S)]
+
+  def generate_final_step_vectors(self, m, step_vectors, strings):
+    storage = {}
+
+    for C in strings:
+      for D in strings:
+        for R in step_vectors:
+          for S in step_vectors:
+            T = [[0] * (m + 1) for _ in range(m + 1)]
+            U = [[0] * (m + 1) for _ in range(m + 1)]
+
+            for i in range(1, m + 1):
+              T[i][0] = R[i]
+              U[0][i] = S[i]
+
+            self.compute_next_step_vectors(C, D, T, U, m)
+
+            R_new = []
+            S_new = []
+
+            for i in range(1, (m + 1)):
+              R_new.append(T[i][m])
+              S_new.append(U[m][i])
+
+            self.store(R_new, S_new, C[1:], D[1:], R[1:], S[1:], storage)
+
+    return storage
+
+  def compute_next_step_vectors(self, C, D, T, U, m):
+    for i in range(1, m + 1):
+      for j in range(1, m + 1):
+        T[i][j] = min(
+            self.substitute_cost_function(C[i], D[j]) - U[i - 1][j],
+            self.delete_cost_function(C[i]),
+            self.insert_cost_function(D[j]) + T[i][j - 1] - U[i - 1][j])
+        U[i][j] = min(
+            self.substitute_cost_function(C[i], D[j]) - T[i][j - 1],
+            self.delete_cost_function(C[i]) + U[i - 1][j] - T[i][j - 1],
+            self.insert_cost_function(D[j])
+        )
+
+  def algorithm_y(self, m, A, step_size_bound):
+    """ Preprocesses data by creating helper submatrices """
+
+    if len(A) == 0:
+      return []
+
+    A = ['#'] + A
+    strings = self.get_all_strings(m, A, '#')
```

Comment:
I'm wondering whether all this handling of '#' may be discarded as it adds somewhat complication to your code. Just in the main function we may use text1[1:] and text2[1:] with helper containing no '#' symbol.